File: Scripts/Blender/Main/BlenderCore.py
# RemakeRegistry\Games\TheSimpsonsGame-PS3\Scripts\Blender\Main\BlenderCore.py
import subprocess
from pathlib import Path
import argparse
import sqlite3
import multiprocessing
import os
import sys
from functools import partial # ADDED: For cleaner worker function calls
from collections import namedtuple # ADDED: For more readable results
from tqdm import tqdm # ADDED: For the progress bar

import tempfile # ADDED: For temporary directories
import shutil   # ADDED: For safely removing directories


# Add Utils path for printer
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..', '..', '..', '..', '..', '.')))
from Engine.Utils.printer import Colours, print

# --- Configuration (Kept global as these are constants for the script's lifetime) ---
current_dir = os.path.dirname(os.path.abspath(__file__))
python_script_path = Path(current_dir).parent / "MainPreinstancedConvert.py"
python_extension_file = Path(current_dir).parent / "PreinstancedImportExtension.py"
blender_exe_path = "Tools/Blender/blender-4.0.2-windows-x64/blender.exe"
DB_FILENAME_DEFAULT = "asset_map.sqlite"

# ADDED: A named tuple for structured, readable results from the worker
ProcessResult = namedtuple("ProcessResult", ["asset_id", "success", "message"])

# --- Worker Function (Modified to capture and print output to console) ---
def run_blender_for_asset(asset_row: dict, export_formats: list, be_verbose: bool, use_debug_sleep: bool) -> ProcessResult:
    """
    Worker function that processes a single asset.
    This version now uses an isolated, temporary directory for the Blender addon.
    This version captures subprocess output and prints it to the console if --verbose is used.
    """
    asset_id = asset_row["identifier"]
    filename = asset_row["filename"]

    temp_addon_dir = tempfile.mkdtemp(prefix="blender_addon_")

    try:
        blend_symlink_path = asset_row["blend_symlink"]
        glb_symlink_path = asset_row["glb_symlink"]
        preinstanced_symlink_path = asset_row["preinstanced_symlink"]

        if not all([filename, blend_symlink_path, glb_symlink_path, preinstanced_symlink_path]):
            return ProcessResult(asset_id, False, "Missing required symlink paths or filename")

        blend_symlink_file = os.path.join(blend_symlink_path, filename + ".blend")
        glb_symlink_file = os.path.join(glb_symlink_path, filename + ".glb")
        fbx_symlink_file = os.path.join(glb_symlink_path, filename + ".fbx")
        preinstanced_symlink_file = os.path.join(preinstanced_symlink_path, filename + ".preinstanced")

        if not os.path.isfile(blend_symlink_file):
            return ProcessResult(asset_id, False, f"Blend symlink not found: {blend_symlink_file}")

        run_blender_flag = False
        if 'glb' in export_formats and not os.path.isfile(glb_symlink_file):
            run_blender_flag = True
        if 'fbx' in export_formats and not os.path.isfile(fbx_symlink_file):
            run_blender_flag = True

        if not run_blender_flag:
            return ProcessResult(asset_id, True, f"Skipped: requested exports already exist for {filename}")

        if not os.path.isfile(preinstanced_symlink_file):
            return ProcessResult(asset_id, False, f"Preinstanced symlink missing: {preinstanced_symlink_file}")

        args = [
            str(blender_exe_path),
            "-b", blend_symlink_file,
            "--python", str(python_script_path),
            "--",
            blend_symlink_file,
            preinstanced_symlink_file,
            glb_symlink_file,
            str(python_extension_file),
            "true" if be_verbose else "false",
            "true" if use_debug_sleep else "false",
            current_dir,
            fbx_symlink_file,
            asset_id,
            temp_addon_dir,
            ",".join(sorted(list(export_formats))),
        ]
        # CHANGE: Conditionally print the output based on the --verbose flag
        if be_verbose:
            # CHANGE: Use subprocess.run to wait for the command and capture its output
            proc = subprocess.run(args, capture_output=True, text=True, encoding='utf-8')

            # Use a lock to prevent interleaved printing from multiple processes
            #with tqdm.get_lock():

            # allow interleaved printing
            print(colour=Colours.DARKGRAY, message=f"\n--- Output for Asset ID: {asset_id} ---")
            # Print stdout if it contains anything
            if proc.stdout:
                print(colour=Colours.GRAY, message=proc.stdout.strip())
            # Print stderr if it contains anything
            if proc.stderr:
                print(colour=Colours.YELLOW, message=proc.stderr.strip())
            print(colour=Colours.DARKGRAY, message=f"--- End of Output for Asset ID: {asset_id} ---\n")
        else:
            proc = subprocess.run(args, capture_output=True, text=True, encoding='utf-8')

        # Check for errors after execution
        if proc.returncode != 0:
            error_details = proc.stderr or proc.stdout or "Blender process produced no output."
            return ProcessResult(asset_id, False, f"Blender exited with code {proc.returncode}. Details: {error_details}")

        # Post-checks
        if 'glb' in export_formats and not os.path.isfile(glb_symlink_file):
            return ProcessResult(asset_id, False, f"GLB file was not created: {glb_symlink_file}")
        if 'fbx' in export_formats and not os.path.isfile(fbx_symlink_file):
            return ProcessResult(asset_id, False, f"FBX file was not created: {fbx_symlink_file}")

        return ProcessResult(asset_id, True, f"Processed successfully: {filename}")

    except Exception as e:
        return ProcessResult(asset_id, False, f"A critical exception occurred in the worker: {str(e)}")
    finally:
        # ADDED: Ensure the temporary directory is always cleaned up
        if os.path.isdir(temp_addon_dir):
            shutil.rmtree(temp_addon_dir)

# --- Main Orchestration ---
def blender_processing(db_path: str, num_workers: int, export_formats, be_verbose: bool, use_debug_sleep: bool) -> None:
    print(colour=Colours.DARKGRAY, message=f"Starting Blender processing with {num_workers} workers...")

    try:
        conn = sqlite3.connect(db_path)
        conn.row_factory = sqlite3.Row
        # Fetch the rows as sqlite3.Row objects first
        rows_from_db = conn.execute("SELECT identifier, filename, preinstanced_symlink, blend_symlink, glb_symlink FROM asset_map").fetchall()
        conn.close()

        # CHANGE: Convert sqlite3.Row objects to standard dictionaries
        assets_to_process = [dict(row) for row in rows_from_db]

    except sqlite3.Error as e:
        print(colour=Colours.RED, message=f"SQLite error: {e}")
        sys.exit(1)

    if not assets_to_process:
        print(colour=Colours.YELLOW, message=f"No assets found in database: {db_path}")
        return

    # Use 'partial' to create a new function with the constant arguments already "baked in".
    # This is a clean way to pass extra, non-changing arguments to a pool worker.
    worker_func = partial(run_blender_for_asset,
                          export_formats=export_formats,
                          be_verbose=be_verbose,
                          use_debug_sleep=use_debug_sleep)

    results = []
    print(colour=Colours.BLUE, message=f"Found {len(assets_to_process)} assets. Dispatching to workers...")

    # Use a multiprocessing Pool and wrap the iterator with tqdm for a progress bar
    with multiprocessing.Pool(processes=num_workers) as pool:
        # pool.imap_unordered is memory efficient and lets us see progress as tasks finish
        result_iterator = pool.imap_unordered(worker_func, assets_to_process)
        for result in tqdm(result_iterator, total=len(assets_to_process), desc="Processing Assets"):
            results.append(result)

    # Now we process the results using the named tuple for clarity
    successes = [r for r in results if r.success]
    failures = [r for r in results if not r.success]
    skipped = [r for r in successes if r.message.startswith("Skipped")]

    print(colour=Colours.GREEN, message=f"\n✅ {len(successes) - len(skipped)} assets processed successfully.")
    print(colour=Colours.YELLOW, message=f"⚪ {len(skipped)} assets skipped (already exist).")
    print(colour=Colours.RED, message=f"❌ {len(failures)} assets failed.")

    if failures:
        print(colour=Colours.RED, message="\n--- Failure Details ---")
        for result in failures:
            print(colour=Colours.RED, message=f"  - ID {result.asset_id}: {result.message}")

def main() -> None:
    parser = argparse.ArgumentParser(description="Process assets in parallel using Blender with an SQLite asset map.")
    parser.add_argument("--verbose", action="store_true", help="Enable verbose output from Blender subprocesses.")
    parser.add_argument("--debug-sleep", action="store_true", help="Enable debug sleep pauses in the Blender script.")
    parser.add_argument("--export", type=str, nargs='*', help="Export formats (e.g., --export fbx glb).")
    parser.add_argument("--db-file-path", type=str, required=True, help=f"Path to SQLite database file.")
    parser.add_argument("--workers", type=int, help="Number of parallel Blender instances. Defaults to number of CPU cores.")
    args = parser.parse_args()

    export_formats = set()
    if args.export:
        for item in args.export:
            export_formats.update(item.lower().split())

    # avoid too many workers if not specified
    if args.workers is None:
        # Calculate 75% of CPU cores, ensuring it's a whole number and at least 1.
        #args.workers = max(1, int(multiprocessing.cpu_count() * 0.75))
        args.workers = int(multiprocessing.cpu_count())

    if args.debug_sleep:
        print(colour=Colours.BLUE, message="Debug sleep mode enabled.")
        debug_sleep = True
    else:
        debug_sleep = False

    if args.verbose:
        print(colour=Colours.BLUE, message="Verbose mode enabled.")
        verbose = True
    else:
        verbose = False

    db_path = os.path.abspath(args.db_file_path) if not os.path.isabs(args.db_file_path) else args.db_file_path

    # --- Path and Argument Validation ---
    for path in [blender_exe_path, python_script_path, python_extension_file, db_path]:
        if not os.path.exists(path):
            print(colour=Colours.RED, message=f"Error: Required file or directory not found: {path}")
            sys.exit(1)

    # delete blend.log before starting new process
    blend_log_path = os.path.join(os.path.dirname(__file__), "blend.log")
    if os.path.exists(blend_log_path):
        os.remove(blend_log_path)

    print(colour=Colours.BLUE, message=f"Export formats: {export_formats or 'None'}")
    print(colour=Colours.BLUE, message=f"Database: {db_path}")
    print(colour=Colours.BLUE, message=f"Workers: {args.workers}")

    blender_processing(db_path, args.workers, export_formats, verbose, debug_sleep)
    print(colour=Colours.GREEN, message="\nProcessing complete.")

if __name__ == "__main__":
    main()

File: operations/Blender/BlenderCore.lua
--[[
BlenderCore.lua – Batch Blender export driver for TSG PS3 assets
================================================================

Overview
    Drives headless Blender to generate exports (GLB/FBX) from prepared
    .blend files using a Python driver and extension. Asset metadata and
    symlink locations are read from an SQLite database (asset_map table).

Environment contracts
    - Global `sdk` table is expected to provide utilities:
        sdk.run_process(cmd, opts)           -- REQUIRED
        sdk.colour_print({colour, message})  -- optional
        sdk.sleep(seconds)                   -- optional
        sdk.ensure_dir(path)                 -- optional
        sdk.path_exists(path)                -- optional
        sdk.is_file(path)                    -- optional
        sdk.remove_file(path), sdk.remove_dir(path) -- optional
    - Global `sqlite` module is expected with:
        sqlite.open(path) -> db
        db.query(sql) -> array of rows
        db.close()

Public entry point
    BlenderCore.main(opts)
        - opts.verbose: boolean
        - opts.debug_sleep: boolean
        - opts.export: boolean
        - opts.export_formats: list of strings (e.g. {"glb", "fbx"})
        - opts.db_file_path: string
        - opts.blender_exe_path: string (path to Blender executable)

Behavior
    1) Validates required files (Blender exe, Python scripts, SQLite DB)
    2) Loads assets from DB
    3) For each asset: checks if requested exports already exist; runs Blender
    4) Summarizes results; throws if any asset failed

Notes
    - Paths are normalized to the host OS separator
    - Missing optional helpers degrade gracefully
    - Subprocess output is always logged; consider gating by verbose in future
]]

---@class Asset
---@field identifier string
---@field filename string
---@field preinstanced_symlink string
---@field blend_symlink string
---@field glb_symlink string

---@class BlenderCoreOptions
---@field verbose boolean|nil
---@field debug_sleep boolean|nil
---@field export boolean|nil
---@field export_formats table<integer, string>|nil
---@field db_file_path string|nil
---@field game_root_path string|nil
---@field preinstanced_dir string|nil
---@field workers integer|nil
---@field batch_size integer|nil

---@class BlenderRunResult
---@field asset_id string
---@field success boolean
---@field skipped boolean
---@field message string|nil

---@class BlenderCore
---@field main fun(opts: BlenderCoreOptions)


if not sqlite then
    error("sqlite module is not available; ensure LuaScriptAction exposes sqlite helpers")
end
-- run_process is always available in engine runtime
if not sdk.run_process then
    error("sdk.run_process helper is required for BlenderCore.lua")
end

local path_sep = package.config:sub(1, 1)

sdk.color_print({ colour = "cyan", message = string.format("[BlenderRun] Detected path separator: '%s'", path_sep) })

---@type BlenderUtils
import(script_dir .. join(path_sep.. "init", "BlenderUtils.lua"))


--- Create and return a temporary directory path for per-run addon files.
-- Uses os.tmpname as a base; attempts to remove the temp file first.
---@param prefix string|nil
---@return string temp directory path
local function make_temp_dir(prefix)
    prefix = prefix or "blender_addon_"
    -- Use a workspace-local temp root to avoid relying on os.tmpname (not available in sandbox)
    local temp_root = join("TMP", "blender_temp")
    sdk.ensure_dir(temp_root)
    local dir = join(temp_root, prefix .. tostring(os.time()) .. "_" .. tostring(math.random(100000, 999999)))
    sdk.ensure_dir(dir)
    return dir
end

-- Seed RNG once (used only for temp dir suffixes)
math.randomseed(os.time())

--TODO make run script pass these paths
-- Repository-relative locations used by this pipeline
local script_root = join("EngineApps", "Games", "TheSimpsonsGame-PS3", "operations")
local blender_dir = join(script_root, "Blender")
local python_script_path = join(blender_dir, "MainPreinstancedConvert.py")
local python_extension_path = join(blender_dir, "blender_addon")
--- Convert a list of export format hints to a set and an ordered list.
-- Recognized tokens: "glb", "fbx" (case-insensitive). Duplicates removed.
---@param list table<integer, string>|nil
---@return table<string, boolean> set
---@return table<integer, string> ordered
local function normalize_export_formats(list)
    local set = {}
    local ordered = {}
    if list then
        for _, value in ipairs(list) do
            local lowered = tostring(value):lower()
            for token in lowered:gmatch("[^%s]+") do
                if token == "glb" or token == "fbx" then
                    if not set[token] then
                        set[token] = true
                        table.insert(ordered, token)
                    end
                end
            end
        end
    end
    return set, ordered
end


--- Load all assets from the SQLite `asset_map` table.
-- Expected columns: identifier, filename, preinstanced_symlink, blend_symlink, glb_symlink
---@param db_path string
---@return table<integer, Asset> array of asset tables
local function load_assets(db_path)
    local db = sqlite.open(to_long_path(db_path))
    local rows = db.query("SELECT identifier, filename, preinstanced_symlink, blend_symlink, glb_symlink FROM asset_map")
    local assets = {}
    for index, row in ipairs(rows) do
        assets[index] = {
            identifier = row.identifier,
            filename = row.filename,
            preinstanced_symlink = row.preinstanced_symlink,
            blend_symlink = row.blend_symlink,
            glb_symlink = row.glb_symlink
        }
    end
    db.close()
    return assets
end

--- Check whether an asset has enough information and whether work is needed.
---@param asset Asset
---@param export_set table<string, boolean> set-like { glb=true?, fbx=true? }
---@return boolean ok, boolean run_needed, string|nil reason
local function should_process_asset(asset, export_set)
    if not asset.blend_symlink or not asset.preinstanced_symlink then
        return false, false, "Missing required symlink paths"
    end

    local blend_file = get_path(asset.blend_symlink, asset.filename, ".blend")
    if not sdk.is_file(blend_file) then
        return false, false, string.format("Blend file not found: %s", blend_file)
    end

    local run_needed = false
    if export_set.glb then
        if not asset.glb_symlink then
            return false, false, "GLB symlink path missing"
        end
        local glb_file = get_path(asset.glb_symlink, asset.filename, ".glb")
        if not sdk.is_file(glb_file) then
            run_needed = true
        end
    end

    if export_set.fbx then
        if not asset.glb_symlink then
            return false, false, "FBX symlink path missing"
        end
        local fbx_file = get_path(asset.glb_symlink, asset.filename, ".fbx")
        if not sdk.is_file(fbx_file) then
            run_needed = true
        end
    end

    return true, run_needed, nil
end

--- Run Blender headless for a single asset if required.
-- Builds the command line for Blender, invokes it, and maps results to
-- a structured record.
---@param asset Asset
---@param export_set table<string, boolean> set-like { glb=true?, fbx=true? }
---@param ordered_formats table<integer, string>
---@param verbose boolean
---@param debug_sleep boolean
---@param blender_exe_path string
---@param Game_Root string
---@param json_db_path string|nil
---@param preinstanced_dir string|nil
---@return BlenderRunResult
local function run_blender_for_asset(asset, export_set, ordered_formats, verbose, debug_sleep, blender_exe_path, Game_Root, json_db_path, preinstanced_dir)
    local ok, run_needed, reason = should_process_asset(asset, export_set)
    if not ok then
        return { asset_id = asset.identifier, success = false, skipped = false, message = reason }
    end

    if not run_needed then
        return { asset_id = asset.identifier, success = true, skipped = true, message = string.format("Skipped: requested exports already exist for %s", asset.filename) }
    end

    -- Informational log for the upcoming Blender run
    log(Colours.DARKCYAN, string.format("running blender for asset %s", tostring(asset.identifier)))

    local blend_file = get_path(asset.blend_symlink, asset.filename, ".blend")
    local glb_file = get_path(asset.glb_symlink, asset.filename, ".glb")
    local fbx_file = get_path(asset.glb_symlink, asset.filename, ".fbx")
    local preinstanced_file = get_path(asset.preinstanced_symlink, asset.filename, ".preinstanced")

    if not sdk.is_file(preinstanced_file) then
        return { asset_id = asset.identifier, success = false, skipped = false, message = string.format("Preinstanced symlink missing: %s", preinstanced_file) }
    end

    -- Create a per-run temp addon directory; will be removed after execution
    local temp_addon_dir = make_temp_dir("blender_addon_")

    local batch_file = join(temp_addon_dir, "batch.json")
    local batch_data = {
        {
            asset_id = asset.identifier,
            blend_file = blend_file,
            preinstanced_file = preinstanced_file,
            glb_file = glb_file,
            fbx_file = fbx_file
        }
    }
    local fh = io.open(batch_file, "w")
    if fh then
        ---@cast fh FileHandle
        fh:write(sdk.text.json.encode(batch_data))
        fh:close()
    else
        log(Colours.RED, "Failed to write batch file: " .. batch_file)
        return { asset_id = asset.identifier, success = false, skipped = false, message = "Failed to write batch file" }
    end

    -- Blender CLI layout: `--` separates Blender's args from script args
    local command = {
        blender_exe_path,
        "-b",
        "--python", python_script_path,
        "--",
        "--batch_file", batch_file,
        "--python_extension_path", python_extension_path,
        "--current_dir", blender_dir,
        "--temp_addon_dir", temp_addon_dir,
        "--game_root_path", Game_Root,
        "--export_formats", table.concat(ordered_formats, ",")
    }

    if verbose then table.insert(command, "--verbose") end
    if debug_sleep then table.insert(command, "--debug_sleep") end
    if json_db_path then
        table.insert(command, "--db_path")
        table.insert(command, json_db_path)
    end
    if preinstanced_dir then
        table.insert(command, "--preinstanced_dir")
        table.insert(command, preinstanced_dir)
    end

    local result
    local status, err = pcall(function()
        -- Capture stdout/stderr for diagnostics and summary reporting
        result = sdk.run_process(command, { capture_stdout = true, capture_stderr = true })
    end)

    if not status then
        if sdk.remove_dir then
            sdk.remove_dir(temp_addon_dir)
        end
        return { asset_id = asset.identifier, success = false, skipped = false, message = string.format("Process launch failed: %s", err) }
    end

    -- Subprocess output is echoed to help debug Blender/Python failures
    log(Colours.DARKGRAY, string.format("\n--- Output for Asset ID: %s ---", asset.identifier))
    if result.stdout and #result.stdout > 0 then
        log(Colours.GRAY, result.stdout)
    end
    if result.stderr and #result.stderr > 0 then
        log(Colours.YELLOW, result.stderr)
    end
    log(Colours.DARKGRAY, string.format("--- End of Output for Asset ID: %s ---\n", asset.identifier))

    if sdk.remove_dir then
        sdk.remove_dir(temp_addon_dir)
    end

    if result.exit_code ~= 0 then
        local message = result.stderr and #result.stderr > 0 and result.stderr:match("[^\n]*") or result.stdout or "Blender process reported failure"
        return { asset_id = asset.identifier, success = false, skipped = false, message = message }
    end

    return { asset_id = asset.identifier, success = true, skipped = false, message = string.format("Processed asset %s", asset.identifier) }
end

local BlenderCore = {}

--- Entry point for batch processing.
-- Validates resources, loads assets, processes sequentially, and summarizes.
-- Throws an error if any asset failed to process.
---@param opts BlenderCoreOptions
---@return nil
function BlenderCore.main(opts)
    opts = opts or {}
    VERBOSE = not not opts.verbose
    local debug_sleep = not not opts.debug_sleep
    local export_set, ordered_formats = normalize_export_formats(opts.export_formats)

    local db_path = opts.db_file_path and normalize(opts.db_file_path) or nil
    if db_path then
        ---@cast db_path string
        local normalized_db_path = normalize(tostring(db_path))
        ---@cast normalized_db_path string
        db_path = to_long_path(normalized_db_path)
    end
    if not db_path or db_path == "" then
        error("DB file path must be specified in opts.db_file_path")
        os.exit(1)
    end

    local json_db_path = opts.preinstanced_dir and join(opts.preinstanced_dir, "normalized_map.json")
    if json_db_path then
        ---@cast json_db_path string
        local normalized_json_db_path = normalize(tostring(json_db_path))
        ---@cast normalized_json_db_path string
        json_db_path = to_long_path(normalized_json_db_path)
    end

    -- use engine tool resolver to get blender
    local blender_exe = nil
    local blender_exe_path = nil
    blender_exe = ResolveToolPath("Blender")
    if blender_exe and blender_exe ~= "" then
        blender_exe_path = normalize(blender_exe)
        ---@cast blender_exe_path string
        log(Colours.CYAN, string.format("Resolved Blender executable via engine tool resolver: %s", tostring(blender_exe_path)), "BlenderRun")
    end


    if blender_exe_path and blender_exe_path ~= "" then
        local normalized_blender_exe_path = normalize(tostring(blender_exe_path))
        ---@cast normalized_blender_exe_path string
        blender_exe_path = to_long_path(normalized_blender_exe_path) or ""
    end

    log(Colours.CYAN, "all input opts: " .. sdk.text.json.encode(opts))

    log(Colours.CYAN, string.format("Export formats: %s", (#ordered_formats > 0) and table.concat(ordered_formats, ", ") or "None"))
    log(Colours.CYAN, string.format("Using DB: %s", db_path))
    log(Colours.CYAN, string.format("Using Blender executable: %s", blender_exe_path))
    if not Game_Root then
        error("Game root path must be specified in opts.Game_Root")
        os.exit(1)
    end
    log(Colours.CYAN, string.format("Using game root path: %s", Game_Root))

    if not sdk.path_exists(blender_exe_path) then
        error(string.format("Blender executable not found: %s", blender_exe_path))
    end
    if not sdk.path_exists(python_script_path) then
        error(string.format("Python driver not found: %s", python_script_path))
    end
    if not sdk.path_exists(python_extension_path) then
        error(string.format("Python extension not found: %s", python_extension_path))
    end
    if not sdk.path_exists(db_path) then
        error(string.format("DB not found: %s", db_path))
    end

    local assets = load_assets(db_path)
    log(Colours.CYAN, string.format("Loaded %d assets from DB", #assets))

    local successes = {}
    local failures = {}
    local skipped = {}

    -- Prepare list of assets that actually require processing
    local work_queue = {}
    for _, asset in ipairs(assets) do
        local ok, need, reason = should_process_asset(asset, export_set)
        if not ok then
            table.insert(failures, { asset_id = asset.identifier, success = false, skipped = false, message = reason })
        elseif not need then
            table.insert(skipped, { asset_id = asset.identifier, success = true, skipped = true, message = string.format("Skipped: requested exports already exist for %s", asset.filename) })
        else
            table.insert(work_queue, asset)
        end
    end

    local p = progress.panel.new(#work_queue, "blender-batch", "Exporting Assets with Blender...")

    local has_spawn = (sdk.spawn_process ~= nil)
    -- Default workers: user-specified opts.workers -> global cpu_count -> sdk.cpu_count -> 1
    local default_workers = 1
    if _G.cpu_count then
        default_workers = tonumber(_G.cpu_count) or default_workers
    elseif sdk and sdk.cpu_count then
        default_workers = tonumber(sdk.cpu_count) or default_workers
    end
    local max_workers = tonumber(opts.workers) or default_workers
    if max_workers < 1 then max_workers = 1 end

    if not has_spawn then
        -- Fallback: run sequentially using existing run_blender_for_asset implementation
        log(Colours.YELLOW, "spawn_process unavailable; running sequentially using sdk.run_process")
        for _, asset in ipairs(work_queue) do
            local rec = run_blender_for_asset(asset, export_set, ordered_formats, VERBOSE, debug_sleep, blender_exe_path, Game_Root, json_db_path, opts.preinstanced_dir)
            if rec.success then
                table.insert(successes, rec)
            else
                table.insert(failures, rec)
            end
            p:Update(1)
        end
    else
        -- Concurrent execution using spawn/poll with JSON batching
        log(Colours.GREEN, string.format("Spawning up to %d workers using sdk.spawn_process", max_workers))

        local active = {} -- pid -> { batch_file=..., temp_dir=..., assets=... }
        local batches = {}

        -- Partition work_queue into discrete batches for parallel processing
        local batch_size = tonumber(opts.batch_size) or 50
        for i = 1, #work_queue, batch_size do
            local batch = {}
            for j = i, math.min(i + batch_size - 1, #work_queue) do
                table.insert(batch, work_queue[j])
            end
            table.insert(batches, batch)
        end

        ---@return integer
        local function active_count()
            local c = 0
            for _ in pairs(active) do c = c + 1 end
            return c
        end

        ---@return nil
        local function start_next()
            if #batches == 0 then return end
            if active_count() >= max_workers then return end

            local batch_assets = table.remove(batches, 1)
            if #batch_assets == 0 then return end

            local temp_addon_dir = make_temp_dir("blender_addon_")
            local batch_file = join(temp_addon_dir, "batch.json")

            local batch_data = {}
            for _, asset in ipairs(batch_assets) do
                ---@cast asset Asset
                local blend_file = get_path(asset.blend_symlink, asset.filename, ".blend")
                local glb_file = get_path(asset.glb_symlink, asset.filename, ".glb")
                local fbx_file = get_path(asset.glb_symlink, asset.filename, ".fbx")
                local preinstanced_file = get_path(asset.preinstanced_symlink, asset.filename, ".preinstanced")

                table.insert(batch_data, {
                    asset_id = asset.identifier,
                    blend_file = blend_file,
                    preinstanced_file = preinstanced_file,
                    glb_file = glb_file,
                    fbx_file = fbx_file
                })
            end

            local fh = io.open(batch_file, "w")
            if fh then
                ---@cast fh FileHandle
                fh:write(sdk.text.json.encode(batch_data))
                fh:close()
            else
                log(Colours.RED, "Failed to write batch file: " .. batch_file)
                return
            end

            local cmd = {
                blender_exe_path,
                "-b",
                "--python", python_script_path,
                "--",
                "--batch_file", batch_file,
                "--python_extension_path", python_extension_path,
                "--current_dir", blender_dir,
                "--temp_addon_dir", temp_addon_dir,
                "--game_root_path", Game_Root,
                "--export_formats", table.concat(ordered_formats, ",")
            }

            if VERBOSE then table.insert(cmd, "--verbose") end
            if debug_sleep then table.insert(cmd, "--debug_sleep") end
            if json_db_path then
                table.insert(cmd, "--db_path")
                table.insert(cmd, json_db_path)
            end
            if opts.preinstanced_dir then
                table.insert(cmd, "--preinstanced_dir")
                table.insert(cmd, opts.preinstanced_dir)
            end

            local ok, res = pcall(function()
                log(Colours.DARKCYAN, string.format("cmd for batch of %d assets: %s", #batch_assets, sdk.text.json.encode(cmd)))
                return sdk.spawn_process(cmd, { capture_stdout = true, capture_stderr = true, cwd = nil })
            end)

            if not ok or not res or not res.pid then
                if sdk.remove_dir then pcall(sdk.remove_dir, temp_addon_dir) end
                local msg = "spawn failed"
                if not ok then msg = tostring(res) end
                for _, asset in ipairs(batch_assets) do
                    ---@cast asset Asset
                    table.insert(failures, { asset_id = asset.identifier, success = false, skipped = false, message = msg })
                    p:Update(1)
                end
                return
            end

            local pid = res.pid
            active[pid] = { batch_file = batch_file, temp_dir = temp_addon_dir, assets = batch_assets }
            log(Colours.DARKCYAN, string.format("Launched PID %s for batch of %d assets", tostring(pid), #batch_assets))
        end

        -- seed initial workers
        for i = 1, max_workers do start_next() end

        -- poll loop
        while next(active) ~= nil or #batches > 0 do
            -- start more if capacity
            while active_count() < max_workers and #batches > 0 do
                start_next()
            end

            -- poll active pids
            for pid, info in pairs(active) do
                local ok, pol = pcall(function() return sdk.poll_process(pid) end)
                if not ok then
                    -- treat as failure and cleanup
                    if sdk.remove_dir then pcall(sdk.remove_dir, info.temp_dir) end
                    for _, asset in ipairs(info.assets) do
                        ---@cast asset Asset
                        table.insert(failures, { asset_id = asset.identifier, success = false, skipped = false, message = "poll failed: " .. tostring(pol) })
                        p:Update(1)
                    end
                    active[pid] = nil
                else
                    -- Process stdout delta for progress markers
                    if pol.stdout_delta and #pol.stdout_delta > 0 then
                        for line in pol.stdout_delta:gmatch("[^\r\n]+") do
                            local asset_id, status, msg = line:match("__REMAKE_ASSET_DONE__:([^:]+):([^:]+):?(.*)")
                            if asset_id and status then
                                if status == "SUCCESS" then
                                    table.insert(successes, { asset_id = asset_id, success = true, skipped = false, message = "Processed asset " .. asset_id })
                                else
                                    table.insert(failures, { asset_id = asset_id, success = false, skipped = false, message = (msg and msg ~= "") and msg or "Failed" })
                                end
                                p:Update(1)
                            else
                                log(Colours.GRAY, string.format("[PID %s] %s", tostring(pid), line))
                            end
                        end
                    end

                    if pol.stderr_delta and #pol.stderr_delta > 0 then
                        for line in pol.stderr_delta:gmatch("[^\r\n]+") do
                            log(Colours.YELLOW, string.format("[PID %s] %s", tostring(pid), line))
                        end
                    end

                    if not pol.running then
                        -- finished
                        if sdk.remove_dir then pcall(sdk.remove_dir, info.temp_dir) end

                        local exit_code = pol.exit_code or 1
                        if exit_code ~= 0 then
                            local message = (pol.stderr and #pol.stderr > 0 and pol.stderr:match("^[^\n]*")) or pol.stdout or "Blender process reported failure"
                            log(Colours.RED, string.format("Batch PID %s failed: %s", tostring(pid), message))
                        else
                            log(Colours.DARKGRAY, string.format("Batch PID %s finished successfully", tostring(pid)))
                        end

                        active[pid] = nil
                    end
                end
            end

            if sdk.sleep then
                pcall(sdk.sleep, 0.05)
            else
                local t0 = os.clock()
                while os.clock() - t0 < 0.05 do end
            end
        end
    end

    p:Complete()

    -- Summarize and error on failures (same behaviour as before)
    log(Colours.CYAN, string.format("Successes: %d, Failures: %d, Skipped: %d", #successes, #failures, #skipped))
    if #failures > 0 then
        log(Colours.RED, "Some assets failed to process. See messages for details.")
        for _, f in ipairs(failures) do
            ---@cast f { asset_id: string, message: string }
            log(Colours.RED, string.format("Asset %s: %s", tostring(f.asset_id), tostring(f.message)))
        end
        error(string.format("%d asset(s) failed to process", #failures))
    end

    log(Colours.GREEN, "\nProcessing complete.")
end

return BlenderCore
